watertap/property_models/activated_sludge/asm1_properties.py

```diff
@@ -282,39 +283,76 @@ def build(self):
+        def material_flow_expression(self, j):
+            if j == "H2O":
+                return self.flow_vol * self.params.dens_mass
+            elif j == "S_ALK":
+                # Convert moles of alkalinity to mass of C assuming all is HCO3-
+                return (
+                    self.flow_vol
+                    * self.alkalinity
+                    * (12 * pyo.units.kg / pyo.units.kmol)
+                )
+            else:
+                return self.flow_vol * self.conc_mass_comp[j]
+
+        self.material_flow_expression = pyo.Expression(
+            self.component_list,
+            rule=material_flow_expression,
+            doc="Material flow terms",
         )
 
-    def get_material_density_terms(b, p, j):
-        if j == "H2O":
-            return b.params.dens_mass
-        elif j == "S_ALK":
-            # Convert moles of alkalinity to mass of C assuming all is HCO3-
-            return b.alkalinity * (12 * pyo.units.kg / pyo.units.kmol)
-        else:
-            return b.conc_mass_comp[j]
+        def enthalpy_flow_expression(self):
+            return (
+                self.flow_vol
+                * self.params.dens_mass
+                * self.params.cp_mass
+                * (self.temperature - self.params.temperature_ref)
+            )
+
+        self.enthalpy_flow_expression = pyo.Expression(
+            rule=enthalpy_flow_expression, doc="Enthalpy flow term"
+        )
+
+        def material_density_expression(self, j):
+            if j == "H2O":
+                return self.params.dens_mass
+            elif j == "S_ALK":
+                # Convert moles of alkalinity to mass of C assuming all is HCO3-
+                return self.alkalinity * (12 * pyo.units.kg / pyo.units.kmol)
+            else:
+                return self.conc_mass_comp[j]
+
+        self.material_density_expression = pyo.Expression(
+            self.component_list,
+            rule=material_density_expression,
+            doc="Material density terms",
+        )
 
-    def get_energy_density_terms(b, p):
-        return (
-            b.params.dens_mass
-            * b.params.cp_mass
-            * (b.temperature - b.params.temperature_ref)
+        def energy_density_expression(self):
+            return (
+                self.params.dens_mass
+                * self.params.cp_mass
+                * (self.temperature - self.params.temperature_ref)
+            )
+
+        self.energy_density_expression = pyo.Expression(
+            rule=energy_density_expression, doc="Energy density term"
         )
 
+    def get_material_flow_terms(self, p, j):
+        return self.material_flow_expression[j]
+
+    def get_enthalpy_flow_terms(self, p):
+        return self.enthalpy_flow_expression
+
+    def get_material_density_terms(self, p, j):
+        return self.material_density_expression[j]
+
+    def get_energy_density_terms(self, p):
+        return self.energy_density_expression
+
```

Comment:
Ah, I see now. You just moved these down and have them returning the new expressions.